c2rust-analyze/src/analyze.rs

```diff
@@ -993,7 +993,10 @@ fn run(tcx: TyCtxt) {
                 if !node_info.unique {
                     perms.remove(PermissionSet::UNIQUE);
                 }
-                // TODO: PermissionSet::NON_NULL
+                if let Some(true) = g.is_null {
+                    // TODO: is this enough?
+                    perms.remove(PermissionSet::NON_NULL);
```

Comment:
Looks right to me.  It's in line with the handling of `UNIQUE`, which is the other forward-flowing permission aside from `NON_NULL`.

Reply:
I was wondering if we want the dynamic analysis to also do the opposite: add `NON_NULL` to pointers where the static algorithm removed it (i.e. we never saw a null pointer in the event log). That might be more useful, but also less sound and harder to implement.

Reply:
Right - I think that's something to add in a separate PR, since it requires some additional changes.